modules/auxiliary/scanner/rdp/cve_2019_0708_bluekeep.rb

```diff
@@ -0,0 +1,564 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'digest'
+require 'rc4'
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Remote::Tcp
+  include Msf::Auxiliary::Scanner
+  include Msf::Auxiliary::Report
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'CVE-2019-0708 BlueKeep Microsoft Remote Desktop RCE Check',
+      'Description'    => %q{
+        This module checks a range of hosts for the CVE-2019-0708 vulnerability
+        by binding the MS_T120 channel outside of its normal slot and sending
+        non-DoS packets which respond differently on patched and vulnerable hosts.
+      },
+      'References'     =>
+        [
+          [ 'CVE', '2019-0708' ],
+          [ 'URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2019-0708' ]
+        ],
+      'Author'         =>
+        [
+          'JaGoTu',
+          'zerosum0x0'
+        ],
+      'License'        => MSF_LICENSE,
+      'Notes'          =>
+        {
+            'Stability'   => [ CRASH_SAFE ],
+            'AKA'         => ['BlueKeep']
+        }
+    ))
+
+    register_options(
+      [
+        OptPort.new('RPORT', [ true, 'Remote port running RDP', 3389 ])
+      ])
+  end
+
+  def report_goods
+    report_vuln(
+      :host         => rhost,
+      :port         => rport,
+      :proto        => 'tcp',
+      :name         => self.name,
+      :info         => 'Behavior indicates a missing Microsoft Windows RDP patch for CVE-2019-0708',
+      :refs         => self.references
+    )
+  end
+
+  def check_rdp
+    # code to check if RDP is open or not
+    vprint_status("Verifying RDP protocol...")
+
+    # send connection
+    #sock.put(connection_request)
+    sock.put(["0300002b26e00000000000436f6f6b69653a206d737473686173683d75736572300d0a0100080000000000"].pack("H*"))
```

Comment:
How much of this can we randomize?  For example, `7573657230` == user0 which makes this fingerprintable.

Reply:
I agree and we did have plans at some point to try and strip down packets. However the fingerprintable part that CAN'T be removed is the MS_T120 channel join, which already has a real Snort/Suricata rule.

Reply:
Yeah, I'm still bummed about the `MS_T120` IOC, but so it goes.

Reply:
As a note @zerosum0x0 we see events detected by our honeypot code that the Suricata rule misses.  I bet there is some space to pad/remove channels that will break their detection. I haven't had a chance to look at the rule yet.

Reply:
I haven't look hard enough at the rule. I know a proper one can be made if it is a broken one though. The same thing happened with MS17-010, the initial rules released were pretty easy to bypass with our MSF modules even though there is deterministic detection.

Reply:
I've added FIXME for other places the username, IP, and hostname are leaked so they can be randomized; I have some things to do right now but will get to it later.